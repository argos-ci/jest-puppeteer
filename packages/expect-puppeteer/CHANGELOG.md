# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [4.4.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v4.3.0...v4.4.0) (2019-12-18)


### Bug Fixes

* fix toDisplayDialog ([ce9df04](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/ce9df04642720a6af23e9b25f1b341bc82eca564))


### Features

* increase peerDependencies to allow Puppeteer v2 ([#289](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/289)) ([62a9b81](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/62a9b819cebf67d7e7b36453b79fba06390585ca))
* **expect-puppeteer:** support frames ([#275](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/275)) ([09703ea](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/09703eaeeeab553e13142153b55030db05611f7c))





# [4.3.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v4.2.0...v4.3.0) (2019-07-14)

**Note:** Version bump only for package expect-puppeteer





## [4.1.1](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v4.1.0...v4.1.1) (2019-04-11)

**Note:** Version bump only for package expect-puppeteer





# [4.1.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v4.0.0...v4.1.0) (2019-03-14)


### Features

* **expect-puppeteer:** add visibility option to toMatchElement ([#208](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/208)) ([46d8ec1](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/46d8ec1))





# [4.0.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v3.9.0...v4.0.0) (2019-02-13)

**Note:** Version bump only for package expect-puppeteer





## [3.5.1](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v3.5.0...v3.5.1) (2018-11-11)

**Note:** Version bump only for package expect-puppeteer





<a name="3.4.0"></a>
# [3.4.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v3.3.1...v3.4.0) (2018-09-24)


### Bug Fixes

* support several instances of Jest in parallel ([#138](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/138)) ([275bc71](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/275bc71))


### Features

* **expect-puppeteer:** Update default options to look at connect object ([1f33ea0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/1f33ea0))
* connect to an already existing instance of Chrome ([#100](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/100)) ([3fcbaf8](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/3fcbaf8))





<a name="3.3.0"></a>
# [3.3.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v3.2.1...v3.3.0) (2018-08-15)

**Note:** Version bump only for package expect-puppeteer





<a name="3.2.0"></a>
# [3.2.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v3.1.0...v3.2.0) (2018-06-17)




**Note:** Version bump only for package expect-puppeteer

<a name="3.1.0"></a>
# [3.1.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v3.0.1...v3.1.0) (2018-06-16)




**Note:** Version bump only for package expect-puppeteer

<a name="3.0.1"></a>
## [3.0.1](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v3.0.0...v3.0.1) (2018-05-03)


### Bug Fixes

* **expect-puppeteer:** use the same behaviour on toMatch and toMatchElement ([784bde8](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/784bde8))




<a name="3.0.0"></a>
# [3.0.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v2.4.0...v3.0.0) (2018-05-03)


### Features

* **expect-puppeteer:** added delay option to element.type for toFill matcher ([#52](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/52)) ([ee39ba9](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/ee39ba9))
* **expect-puppeteer:** enhance toMatchElement / toClick text option ([cee8f46](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/cee8f46)), closes [#51](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/51) [#50](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/50)


### BREAKING CHANGES

* **expect-puppeteer:** Text is now trimmed and no longer evaluated as a RegExp. If you want this behaviour, use a true RegExp.




<a name="2.4.0"></a>
# [2.4.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v2.3.0...v2.4.0) (2018-04-24)


### Features

* **expect-puppeteer:** keep original error message ([#45](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/45)) ([72b60a9](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/72b60a9)), closes [#33](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/33)
* **expect-puppeteer:** make default options configurable ([#46](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/46)) ([0d493c4](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/0d493c4))
* adjust default timeout with slowMo ([6871ec8](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/6871ec8)), closes [#36](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/36)




<a name="2.3.0"></a>
# [2.3.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v2.2.3...v2.3.0) (2018-04-06)




**Note:** Version bump only for package expect-puppeteer

<a name="2.2.2"></a>
## [2.2.2](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v2.2.1...v2.2.2) (2018-04-01)


### Bug Fixes

* **expect-puppeteer:** reduce chance of collision ([83ade43](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/83ade43)), closes [#issuecomment-377698539](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/issuecomment-377698539)




<a name="2.2.1"></a>
## [2.2.1](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v2.2.0...v2.2.1) (2018-03-25)


### Bug Fixes

* support original Jest matchers ([99d542e](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/commit/99d542e)), closes [#17](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/issues/17)




<a name="2.2.0"></a>
# [2.2.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer/compare/v2.1.0...v2.2.0) (2018-03-17)




**Note:** Version bump only for package expect-puppeteer

<a name="2.1.0"></a>
# [2.1.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/puppeteer-expect/compare/v2.0.1...v2.1.0) (2018-03-16)


### Features

* add element handle support ([4d37d5b](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/puppeteer-expect/commit/4d37d5b))




<a name="2.0.1"></a>
## [2.0.1](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/puppeteer-expect/compare/v2.0.0...v2.0.1) (2018-03-08)


### Bug Fixes

* **expect-puppeteer:** fix expectMatch when body is undefined ([#9](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/puppeteer-expect/issues/9)) ([0c60970](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/puppeteer-expect/commit/0c60970))




<a name="2.0.0"></a>
# [2.0.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/puppeteer-expect/compare/v1.1.1...v2.0.0) (2018-03-05)


### Features

* simplify expect usage ([e0fc3d1](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/puppeteer-expect/commit/e0fc3d1))


### BREAKING CHANGES

* `expectPage()` is replaced by `expect(page)`




<a name="1.1.1"></a>
## [1.1.1](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/puppeteer-expect/compare/v1.1.0...v1.1.1) (2018-03-04)


### Bug Fixes

* **expect-puppeteer:** add all sources in pkg ([bb51870](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/puppeteer-expect/commit/bb51870))




<a name="1.1.0"></a>
# [1.1.0](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/puppeteer-expect/compare/v1.0.1...v1.1.0) (2018-03-04)


### Features

* add spawnd & expect-puppeteer ([6b7f5a4](https://github.com/smooth-code/jest-puppeteer/tree/master/packages/puppeteer-expect/commit/6b7f5a4))
