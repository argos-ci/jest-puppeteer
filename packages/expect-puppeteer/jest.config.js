module.exports = {
  preset: "<rootDir>/../jest-puppeteer",
  transform: {
    "^.+\\.(t|j)sx?$": ["@swc/jest"],
  },
  testTimeout: 10000,
};
