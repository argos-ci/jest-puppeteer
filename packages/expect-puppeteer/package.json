{
  "name": "expect-puppeteer",
  "description": "Assertion toolkit for Puppeteer.",
  "version": "8.0.5",
  "type": "commonjs",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./setup": "./dist/setup.js",
    "./package.json": "./package.json"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/argos-ci/jest-puppeteer.git",
    "directory": "packages/expect-puppeteer"
  },
  "homepage": "https://github.com/argos-ci/jest-puppeteer/tree/main/packages/expect-puppeteer#readme",
  "bugs": {
    "url": "https://github.com/argos-ci/jest-puppeteer/issues"
  },
  "engines": {
    "node": ">=16"
  },
  "keywords": [
    "jest",
    "puppeteer",
    "jest-puppeteer",
    "chromeless",
    "chrome-headless",
    "expect",
    "assert",
    "should",
    "assertion"
  ],
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "prebuild": "rm -rf dist",
    "build": "rollup -c"
  },
  "devDependencies": {
    "puppeteer": "^20.3.0",
    "rollup": "^3.23.0",
    "rollup-plugin-dts": "^5.3.0",
    "rollup-plugin-swc3": "^0.8.1"
  }
}
