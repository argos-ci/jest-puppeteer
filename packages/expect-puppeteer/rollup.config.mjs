import { swc, defineRollupSwcOption } from "rollup-plugin-swc3";
import dts from "rollup-plugin-dts";

const bundle = (config) => ({
  input: "src/index.ts",
  external: (id) => {
    return !/^[./]/.test(id) || id === "./index.js";
  },
  ...config,
});

const swcPlugin = swc(
  defineRollupSwcOption({
    jsc: {
      parser: {
        syntax: "typescript",
      },
    },
    env: {
      targets: "node 16",
    },
  })
);

export default [
  bundle({
    output: {
      file: "dist/index.js",
      format: "cjs",
    },
    plugins: [swcPlugin],
  }),
  bundle({
    plugins: [dts()],
    output: {
      file: "dist/index.d.ts",
      format: "es",
    },
  }),
];
