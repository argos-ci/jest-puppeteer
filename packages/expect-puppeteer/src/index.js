/* eslint-disable no-use-before-define, no-restricted-syntax, no-await-in-loop */
import { getPuppeteerType } from './utils'
import notToMatch from './matchers/notToMatch'
import notToMatchElement from './matchers/notToMatchElement'
import toClick from './matchers/toClick'
import toDisplayDialog from './matchers/toDisplayDialog'
import toFill from './matchers/toFill'
import toFillForm from './matchers/toFillForm'
import toMatch from './matchers/toMatch'
import toMatchElement from './matchers/toMatchElement'
import toSelect from './matchers/toSelect'
import toUploadFile from './matchers/toUploadFile'
import { setDefaultOptions, getDefaultOptions } from './options'

const pageMatchers = {
  toClick,
  toDisplayDialog,
  toFill,
  toFillForm,
  toMatch,
  toMatchElement,
  toSelect,
  toUploadFile,
  not: {
    toMatch: notToMatch,
    toMatchElement: notToMatchElement,
  },
}

const elementHandleMatchers = {
  toClick,
  toFill,
  toFillForm,
  toMatch,
  toMatchElement,
  toSelect,
  toUploadFile,
  not: {
    toMatch: notToMatch,
    toMatchElement: notToMatchElement,
  },
}

function createMatcher(matcher, page) {
  return async function throwingMatcher(...args) {
    if (typeof global.expect !== 'undefined') {
      global.expect.getState().assertionCalls += 1
    }

    try {
      return await matcher(page, ...args)
    } catch (error) {
      Error.captureStackTrace(error, createMatcher)
      throw error
    }
  }
}

function internalExpect(type, matchers) {
  const expectation = {
    not: {},
  }

  Object.keys(matchers).forEach(key => {
    if (key === 'not') return
    expectation[key] = createMatcher(matchers[key], type)
  })

  Object.keys(matchers.not).forEach(key => {
    expectation.not[key] = createMatcher(matchers.not[key], type)
  })

  return expectation
}

function expectPuppeteer(actual) {
  const type = getPuppeteerType(actual)
  switch (type) {
    case 'Page':
      return internalExpect(actual, pageMatchers)
    case 'ElementHandle':
      return internalExpect(actual, elementHandleMatchers)
    default:
      throw new Error(`${actual} is not supported`)
  }
}

if (typeof global.expect !== 'undefined') {
  const originalExpect = global.expect
  global.expect = (actual, ...args) => {
    const type = getPuppeteerType(actual)
    if (type) {
      const matchers = expectPuppeteer(actual)
      const jestMatchers = originalExpect(actual, ...args)
      return {
        ...jestMatchers,
        ...matchers,
        not: {
          ...jestMatchers.not,
          ...matchers.not,
        },
      }
    }
    return originalExpect(actual, ...args)
  }
  Object.keys(originalExpect).forEach(prop => {
    global.expect[prop] = originalExpect[prop]
  })
}

module.exports = expectPuppeteer
module.exports.setDefaultOptions = setDefaultOptions
module.exports.getDefaultOptions = getDefaultOptions
