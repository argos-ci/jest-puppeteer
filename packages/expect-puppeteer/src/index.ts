// import modules
import { notToMatchTextContent } from "./matchers/notToMatchTextContent";
import { notToMatchElement } from "./matchers/notToMatchElement";
import { toClick } from "./matchers/toClick";
import { toDisplayDialog } from "./matchers/toDisplayDialog";
import { toFill } from "./matchers/toFill";
import { toFillForm } from "./matchers/toFillForm";
import { toMatchTextContent } from "./matchers/toMatchTextContent";
import { toMatchElement } from "./matchers/toMatchElement";
import { toSelect } from "./matchers/toSelect";
import { toUploadFile } from "./matchers/toUploadFile";
import {
  checkIsPuppeteerInstance,
  checkIsElementHandle,
  checkIsFrame,
  checkIsPage,
} from "./utils";

// import interfaces and types
import type { JestExpect } from "@jest/expect";
import type { ElementHandle, Frame, Page } from "puppeteer";
import type { PuppeteerInstance } from "./utils";

// reexport
export { setDefaultOptions, getDefaultOptions } from "./options";

// ---------------------------

// declare native matcher function signature
type PuppeteerMatcher<T> = (page: T, ...args: unknown[]) => Promise<unknown>;

// return intersection type from union type
type Intersection<T> = (T extends unknown ? (k: T) => void : never) extends (
  k: infer R,
) => void
  ? R
  : never;

// declare wrapped matcher function signature
type Wrapper<T> = T extends (
  page: Intersection<PuppeteerInstance>,
  ...args: infer A
) => infer R
  ? (...args: A) => R
  : never;

// declare matchers list
type PuppeteerMatchers<T> = T extends PuppeteerInstance
  ? {
      // common
      toClick: Wrapper<typeof toClick>;
      toFill: Wrapper<typeof toFill>;
      toFillForm: Wrapper<typeof toFillForm>;
      toMatchTextContent: Wrapper<typeof toMatchTextContent>;
      toMatchElement: Wrapper<typeof toMatchElement>;
      toSelect: Wrapper<typeof toSelect>;
      toUploadFile: Wrapper<typeof toUploadFile>;
      // inverse matchers
      not: {
        toMatchTextContent: Wrapper<typeof notToMatchTextContent>;
        toMatchElement: Wrapper<typeof notToMatchElement>;
      };
    }
  : never;

// declare page matchers list
interface PageMatchers extends PuppeteerMatchers<Page> {
  // instance specific
  toDisplayDialog: Wrapper<typeof toDisplayDialog>;
  // inverse matchers
  not: PuppeteerMatchers<Page>[`not`] & {};
}

// declare frame matchers list
interface FrameMatchers extends PuppeteerMatchers<Frame> {
  // inverse matchers
  not: PuppeteerMatchers<Frame>[`not`] & {};
}

// declare element matchers list
interface ElementHandleMatchers
  extends PuppeteerMatchers<ElementHandle<Element>> {
  // inverse matchers
  not: PuppeteerMatchers<ElementHandle<Element>>[`not`] & {};
}

// declare matchers per instance type
type PMatchersPerType<T> = T extends Page
  ? PageMatchers
  : T extends Frame
    ? FrameMatchers
    : T extends ElementHandle<Element>
      ? ElementHandleMatchers
      : never;

// constraint current expect to support puppeteer matchers
export interface PuppeteerExpect {
  <T>(actual: T): PMatchersPerType<T>;
}

// global object signature
type GlobalWithExpect = typeof globalThis & { expect: PuppeteerExpect };

// ---------------------------

// extend global jest object
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    interface Matchers<R, T> {
      // common
      toClick: T extends PuppeteerInstance ? Wrapper<typeof toClick> : never;
      toFill: T extends PuppeteerInstance ? Wrapper<typeof toFill> : never;
      toFillForm: T extends PuppeteerInstance
        ? Wrapper<typeof toFillForm>
        : never;
      toMatchTextContent: T extends PuppeteerInstance
        ? Wrapper<typeof toMatchTextContent>
        : never;
      toMatchElement: T extends PuppeteerInstance
        ? Wrapper<typeof toMatchElement>
        : never;
      toSelect: T extends PuppeteerInstance ? Wrapper<typeof toSelect> : never;
      toUploadFile: T extends PuppeteerInstance
        ? Wrapper<typeof toUploadFile>
        : never;
      // page
      toDisplayDialog: T extends Page ? Wrapper<typeof toDisplayDialog> : never;
      // inverse matchers
      not: {
        toMatchTextContent: T extends PuppeteerInstance
          ? Wrapper<typeof notToMatchTextContent>
          : never;
        toMatchElement: T extends PuppeteerInstance
          ? Wrapper<typeof notToMatchElement>
          : never;
      };
    }
  }
}

// ---------------------------
// @ts-expect-error global node object w/ initial jest expect prop attached
const jestExpect = global.expect as JestExpect;

// ---------------------------
// wrapper executing the matcher and capturing the stack trace on error before rethrowing
const wrapMatcher = <T extends PuppeteerInstance>(
  matcher: PuppeteerMatcher<T>,
  instance: T,
) =>
  async function throwingMatcher(...args: unknown[]): Promise<unknown> {
    // ???
    jestExpect.getState().assertionCalls += 1;
    try {
      // run async matcher
      const result = await matcher(instance, ...args);
      // resolve
      return result;
    } catch (err: unknown) {
      if (err instanceof Error) Error.captureStackTrace(err, throwingMatcher);
      // reject
      throw err;
    }
  } as Wrapper<PuppeteerMatcher<T>>;

// ---------------------------
// create the generic expect object and bind wrapped matchers to it
const puppeteerExpect = <T extends PuppeteerInstance>(instance: T) => {
  // read instance type
  const [isPage, isFrame, isHandle] = [
    checkIsPage(instance),
    checkIsFrame(instance),
    checkIsElementHandle(instance),
  ];

  if (!isPage && !isFrame && !isHandle)
    throw new Error(`${instance} is not supported`);

  // retrieve matchers
  const expectation = {
    // common
    toClick: wrapMatcher(toClick as PuppeteerMatcher<T>, instance),
    toFill: wrapMatcher(toFill as PuppeteerMatcher<T>, instance),
    toFillForm: wrapMatcher(toFillForm as PuppeteerMatcher<T>, instance),
    toMatchTextContent: wrapMatcher(
      toMatchTextContent as PuppeteerMatcher<T>,
      instance,
    ),
    toMatchElement: wrapMatcher(
      toMatchElement as PuppeteerMatcher<T>,
      instance,
    ),
    toSelect: wrapMatcher(toSelect as PuppeteerMatcher<T>, instance),
    toUploadFile: wrapMatcher(toUploadFile as PuppeteerMatcher<T>, instance),
    // page
    toDisplayDialog: isPage
      ? wrapMatcher(toDisplayDialog as PuppeteerMatcher<T>, instance)
      : undefined,
    // inverse matchers
    not: {
      toMatchTextContent: wrapMatcher(
        notToMatchTextContent as PuppeteerMatcher<T>,
        instance,
      ),
      toMatchElement: wrapMatcher(
        notToMatchElement as PuppeteerMatcher<T>,
        instance,
      ),
    },
  };

  return expectation as unknown as PMatchersPerType<T>;
};

// ---------------------------
// merge puppeteer matchers w/ jest matchers and return a new object
const expectPuppeteer = (<T>(actual: T) => {
  // puppeteer
  if (checkIsPuppeteerInstance(actual)) {
    const matchers = puppeteerExpect(actual);
    const jestMatchers = jestExpect(actual);
    return {
      ...jestMatchers,
      ...matchers,
      not: {
        ...jestMatchers.not,
        ...matchers.not,
      },
    };
  }

  // not puppeteer (fall back to jest defaults, puppeteer matchers not available)
  return jestExpect(actual);
}) as PuppeteerExpect & JestExpect;

Object.keys(jestExpect).forEach((prop) => {
  // @ts-expect-error add jest expect properties to expect-puppeteer implementation
  expectPuppeteer[prop] = jestExpect[prop];
});

export { expectPuppeteer as expect };

// replace jest expect by expect-puppeteer ...
if (typeof (global as GlobalWithExpect).expect !== `undefined`)
  (global as GlobalWithExpect).expect = expectPuppeteer;
