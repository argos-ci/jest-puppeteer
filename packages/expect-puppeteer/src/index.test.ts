import { getDefaultOptions, setDefaultOptions } from "expect-puppeteer";

// import globals
import "jest-puppeteer";

expect.addSnapshotSerializer({
  print: () => "hello",
  test: () => true,
  serialize: () => "hello",
});

describe("expect-puppeteer", () => {
  beforeEach(async () => {
    await page.goto(`http://localhost:${process.env.TEST_SERVER_PORT}`);
  });

  it("should work with original Jest matchers", async () => {
    expect(page).toBeDefined();
    expect(page).not.toBe(null);

    const main = await page.$("main");
    expect(main).toBeDefined();
    expect(main).not.toBe(null);

    expect(200).toBe(200);
  });

  it("should works with `addSnapshotSerializer`", () => {
    expect({ hello: "world" }).toMatchSnapshot();
  });

  it("should get and set default options", () => {
    expect(getDefaultOptions()).toEqual({ timeout: 500 });
    setDefaultOptions({ timeout: 200 });
    expect(getDefaultOptions()).toEqual({ timeout: 200 });
  });
});
