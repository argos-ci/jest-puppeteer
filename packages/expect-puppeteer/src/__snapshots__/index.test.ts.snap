// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect-puppeteer should works with \`addSnapshotSerializer\` 1`] = `hello`;
